packages/frontend/app/routes/payments.incoming.$incomingPaymentId.tsx

```diff
@@ -0,0 +1,188 @@
+import { badgeColorByState, formatAmount, prettify } from '~/shared/utils'
+
+export async function loader({ params }: LoaderArgs) {
+  const incomingPaymentId = params.incomingPaymentId
+
+  const result = z.string().uuid().safeParse(incomingPaymentId)
+  if (!result.success) {
+    throw json(null, {
+      status: 400,
+      statusText: 'Invalid incoming payment ID.'
+    })
+  }
+
+  const incomingPayment = await getIncomingPayment({ id: result.data })
+
+  if (!incomingPayment) {
+    throw json(null, { status: 400, statusText: 'Incoming payment not found.' })
+  }
+
+  return json({
+    incomingPayment: {
+      ...incomingPayment,
+      createdAt: new Date(incomingPayment.createdAt).toLocaleString(),
+      expiresAt: new Date(incomingPayment.expiresAt).toLocaleString()
+    }
+  })
+}
+
+export default function ViewIncomingPaymentPage() {
+  const { incomingPayment } = useLoaderData<typeof loader>()
+  const [showMetadata, setShowMetadata] = useState(false)
+
+  return (
+    <div className='pt-4 flex flex-col space-y-4'>
+      <div className='flex flex-col rounded-md bg-offwhite px-6'>
+        {/* Incoming Payment General Info */}
+        <PageHeader className='!justify-end'>
+          <Button aria-label='go back to payments page' to='/payments'>
+            Go to payments page
+          </Button>
+        </PageHeader>
+        <div className='grid grid-cols-1 py-3 gap-6 md:grid-cols-3 border-b border-pearl'>
+          {/* Incoming Payment General Info*/}
+          <div className='col-span-1 pt-3'>
+            <h3 className='text-lg font-medium'>General Information</h3>
+            <p className='text-sm'>
+              Created at {incomingPayment.createdAt}{' '}
+              <Badge color={badgeColorByState[incomingPayment.state]}>
+                {incomingPayment.state}
+              </Badge>
+            </p>
+            {new Date(incomingPayment.expiresAt) > new Date() && (
+              <p className='text-sm mb-2'>
+                Expires at {incomingPayment.expiresAt}{' '}
+              </p>
+            )}
+          </div>
+          <div className='md:col-span-2 bg-white rounded-md shadow-md'>
+            <div className='w-full p-4 space-y-3'>
+              <div>
+                <p className='font-medium'>Incoming Payment ID</p>
+                <p className='mt-1'>{incomingPayment.id}</p>
+              </div>
+              <div>
+                <p className='font-medium'>
+                  Wallet Address ID{' '}
+                  <Button
+                    className='mt-1 ml-1'
+                    aria-label='go to wallet address page'
+                    type='button'
+                    size='xs'
+                    to={`/wallet-addresses/${incomingPayment.walletAddressId}`}
+                  >
+                    ↗️
+                  </Button>
+                </p>
+                <p className='mt-1'>{incomingPayment.walletAddressId}</p>
```

Comment:
Thoughts about this approach? The same can be applied on outgoing payments.
```suggestion
               <div>
                <p className='font-medium'>Wallet Address ID </p>
                <Link
                  to={`/wallet-addresses/${incomingPayment.walletAddressId}`}
                  className='mt-1 underline text-blue-500'
                >
                  {incomingPayment.walletAddressId}
                </Link>
              </div>
```

Reply:
I considered this as well. I think this is probably better. I committed this + some styling for `hover` and `visited`.